plugins/cffi/cffi_asyncio.py

```diff
@@ -0,0 +1,854 @@
+                    else:
+                        print("surprise opcode", opcode)
+                    yield value
+                else:
+                    print("no msg", wsgi_req.websocket_opcode)
+                    await event.wait()
+                    event.clear()
+
+            # { "type": "websocket.disconnect", "code": int }
+
+        receive = receiver().__anext__
+
+        async def send(event):
+            print("ws send", event)
+            if event["type"] == "websocket.accept":
+                if (
+                    lib.uwsgi_websocket_handshake(
+                        wsgi_req, ffi.NULL, 0, ffi.NULL, 0, ffi.NULL, 0
+                    )
+                    < 0
+                ):
+                    raise IOError("unable to send websocket handshake")
+
+            elif event["type"] == "websocket.send":
+                # ok to call during any part of app?
+                try:
+                    msg = event["bytes"]
+                    if (
+                        lib.uwsgi_websocket_send_binary(
+                            wsgi_req, ffi.new("char[]", msg), len(msg)
+                        )
+                        < 0
+                    ):
+                        raise IOError("unable to send websocket message")
+                except KeyError:
+                    msg = event["text"].encode("utf-8")
+                    if (
+                        lib.uwsgi_websocket_send(
+                            wsgi_req, ffi.new("char[]", msg), len(msg)
+                        )
+                        < 0
+                    ):
+                        raise IOError("unable to send websocket message")
+
+            elif event["type"] == "websocket.close":
+                gc.switch(event)
+
+    elif scope["type"] == "http":
+
+        async def receive():
+            return {"type": "http.request"}
+
+    app_task = loop.create_task(app(scope, receive, send))
+
+    while True:
+        event = gc.parent.switch()
+        if event["type"] == "http.response.start":
+            # raw uwsgi function accepts bytes
+            asgi_start_response(wsgi_req, event["status"], event["headers"])
+        elif event["type"] == "http.response.body":
+            data = event["body"]
+            print(
+                "write_body_do",
+                lib.uwsgi_response_write_body_do(
+                    wsgi_req, ffi.new("char[]", data), len(data)
+                ),
+            )
+            if not event.get("more_body"):
+                break
+        elif event["type"] == "websocket.close":
+            break
+        else:
+            print("loop event", event)
+
+    return lib.UWSGI_OK
+
+
+def to_network(native):
+    return native.encode("latin1")
+
+
+def iscoroutine(app):
+    """
+    Could app be ASGI?
+    """
+    return inspect.iscoroutinefunction(app) or inspect.iscoroutinefunction(app.__call__)
+
+
+ASGI_CALLABLE = ffi.cast("void *", 2)
+
+
+@ffi.def_extern()
```

Comment:
Here we redefine the request callback compared to the other part of the plugin `cffi_init.py`